### ISSUE TITLE
perl segfaults

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6089]
Migrated from [rt.perl.org#18579](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=18579) (status was 'resolved')

Searchable as RT18579$

**Comment by p5pRT** [comment]
### From runefro@ulrik.uio.no

#### Created by runefro@ulrik.uio.no



Running the below program results in&#8203;:

&nbsp;  Attempt to free unreferenced scalar at \.\./perl\-bug\-segfault line 10\.
&nbsp;  Segmentation fault

Uncommenting the find line\, and commenting the @&#8203;files declaration
segfaults with no message\.

\#\!/usr/bin/perl                                                                 
use File&#8203;::Find;

sub wanted\{
&nbsp;    push\(@&#8203;files\, $File&#8203;::Find&#8203;::name\);
\}

\#find\(\\&wanted\, '\.'\);                                                           
@&#8203;files = \('foo'\, 'ar'\, 'gazonk'\);
for my $path \(sort\(@&#8203;files\)\) \{
&nbsp;    $t = '\\0\\A';
&nbsp;    $t =~ s/\\\\\(\[nA\]\)/&convpart\($\`\, $1\)/eg;
\}

sub convpart \{
&nbsp;    my \($dir\, $fname\) = \( 'foo' =~ m\,\(\.\*/\)\(\[^/\]\+\)$\,\);
&nbsp;    opendir\(DIR\, $dir\);
&nbsp;    @&#8203;files = sort\(grep \{/^$fname/\} readdir\(DIR\)\);
&nbsp;    closedir\(DIR\);
\}

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.8.0:

Configured by bhcompile at Sun Sep  1 23:55:07 EDT 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.18-11smp, archname=i386-linux-thread-multi
    uname='linux daffy.perf.redhat.com 2.4.18-11smp #1 smp thu aug 15 06:41:59 edt 2002 i686 i686 i386 gnulinux '
    config_args='-des -Doptimize=-O2 -march=i386 -mcpu=i686 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dinstallprefix=/usr -Dprefix=/usr -Darchname=i386-linux -Dvendorprefix=/usr -Dsiteprefix=/usr -Duseshrplib -Dusethreads -Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -march=i386 -mcpu=i686',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -I/usr/include/gdbm'
    ccversion='', gccversion='3.2 20020822 (Red Hat Linux Rawhide 3.2-5)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.92.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.2.92'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.8.0/i386-linux-thread-multi/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/lib/perl5/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.0
    /usr/lib/perl5/vendor_perl
    .


Environment for perl v5.8.0:
    HOME=/home/runefro
    LANG=en_US.iso885915
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/local/runefro/gnu/bin:/home/runefro/bin/linux:/home/runefro/bin:/local2/bin:/usr/bin:/usr/X11R6/bin:/local/gnu/bin:/local/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/home/runefro/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From rune.froysa@usit.uio.no

The bug in question is a duplicate\.  Sendmail decided to transfer some
mail that I thought was lost\.  I guess I should have remembered to
check the mail queue before starting it\.

/Rune

**Comment by p5pRT** [comment]
@rspier - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Nov 21\, 2002 at 10&#8203;:20&#8203;:32PM \-0000\, runefro@&#8203;ulrik\.uio\.no \(via RT\) wrote&#8203;:
> Running the below program results in&#8203;:
> 
>   Attempt to free unreferenced scalar at \.\./perl\-bug\-segfault line 10\.
>   Segmentation fault

The code below can reduced the simpler

&nbsp;    @&#8203;a = \(1\,2\,3\);
&nbsp;    for \(sort\(@&#8203;a\)\) \{ 
&nbsp;    @&#8203;a = \(\);
&nbsp;    \}

which is another variant on the 'modififying the array being iterated
causes strange things to happen' bug that gets reported regularly\. 

The workaround in your case is to avoid modifying @&#8203;files within the for
loop; in the longer term\, Perl should be fixed so that it doesn't coredump
in such situations \(but this is hard to do efficiently\)\.

> Uncommenting the find line\, and commenting the @&#8203;files declaration
> segfaults with no message\.
> 
> \#\!/usr/bin/perl                                                                 
> use File&#8203;::Find;
> 
> sub wanted\{
>     push\(@&#8203;files\, $File&#8203;::Find&#8203;::name\);
> \}
> 
> \#find\(\\&wanted\, '\.'\);                                                           
> @&#8203;files = \('foo'\, 'ar'\, 'gazonk'\);
> for my $path \(sort\(@&#8203;files\)\) \{
>     $t = '\\0\\A';
>     $t =~ s/\\\\\(\[nA\]\)/&convpart\($\`\, $1\)/eg;
> \}
> 
> sub convpart \{
>     my \($dir\, $fname\) = \( 'foo' =~ m\,\(\.\*/\)\(\[^/\]\+\)$\,\);
>     opendir\(DIR\, $dir\);
>     @&#8203;files = sort\(grep \{/^$fname/\} readdir\(DIR\)\);
>     closedir\(DIR\);
> \}
> 
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=high
> \-\-\-
> Site configuration information for perl v5\.8\.0&#8203;:
> 
> Configured by bhcompile at Sun Sep  1 23&#8203;:55&#8203;:07 EDT 2002\.
> 
> Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
>   Platform&#8203;:
>     osname=linux\, osvers=2\.4\.18\-11smp\, archname=i386\-linux\-thread\-multi
>     uname='linux daffy\.perf\.redhat\.com 2\.4\.18\-11smp \#1 smp thu aug 15 06&#8203;:41&#8203;:59 edt 2002 i686 i686 i386 gnulinux '
>     config\_args='\-des \-Doptimize=\-O2 \-march=i386 \-mcpu=i686 \-Dmyhostname=localhost \-Dperladmin=root@&#8203;localhost \-Dcc=gcc \-Dcf\_by=Red Hat\, Inc\. \-Dinstallprefix=/usr \-Dprefix=/usr \-Darchname=i386\-linux \-Dvendorprefix=/usr \-Dsiteprefix=/usr \-Duseshrplib \-Dusethreads \-Duseithreads \-Duselargefiles \-Dd\_dosuid \-Dd\_semctl\_semun \-Di\_db \-Ui\_ndbm \-Di\_gdbm \-Di\_shadow \-Di\_syslog \-Dman3ext=3pm \-Duseperlio \-Dinstallusrbinperl \-Ubincompat5005 \-Uversiononly \-Dpager=/usr/bin/less \-isr'
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
>     useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>     use64bitint=undef use64bitall=undef uselongdouble=undef
>     usemymalloc=n\, bincompat5005=undef
>   Compiler&#8203;:
>     cc='gcc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-I/usr/include/gdbm'\,
>     optimize='\-O2 \-march=i386 \-mcpu=i686'\,
>     cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing \-I/usr/include/gdbm'
>     ccversion=''\, gccversion='3\.2 20020822 \(Red Hat Linux Rawhide 3\.2\-5\)'\, gccosandvers=''
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
>     alignbytes=4\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='gcc'\, ldflags =' \-L/usr/local/lib'
>     libpth=/usr/local/lib /lib /usr/lib
>     libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt \-lutil
>     perllibs=\-lnsl \-ldl \-lm \-lpthread \-lc \-lcrypt \-lutil
>     libc=/lib/libc\-2\.2\.92\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
>     gnulibc\_version='2\.2\.92'
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic \-Wl\,\-rpath\,/usr/lib/perl5/5\.8\.0/i386\-linux\-thread\-multi/CORE'
>     cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'
> 
> Locally applied patches&#8203;:
>     
> 
> \-\-\-
> @&#8203;INC for perl v5\.8\.0&#8203;:
>     /usr/lib/perl5/5\.8\.0/i386\-linux\-thread\-multi
>     /usr/lib/perl5/5\.8\.0
>     /usr/lib/perl5/site\_perl/5\.8\.0/i386\-linux\-thread\-multi
>     /usr/lib/perl5/site\_perl/5\.8\.0
>     /usr/lib/perl5/site\_perl
>     /usr/lib/perl5/vendor\_perl/5\.8\.0/i386\-linux\-thread\-multi
>     /usr/lib/perl5/vendor\_perl/5\.8\.0
>     /usr/lib/perl5/vendor\_perl
>     \.
> 
> \-\-\-
> Environment for perl v5\.8\.0&#8203;:
>     HOME=/home/runefro
>     LANG=en\_US\.iso885915
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>     PATH=/local/runefro/gnu/bin&#8203;:/home/runefro/bin/linux&#8203;:/home/runefro/bin&#8203;:/local2/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:/local/gnu/bin&#8203;:/local/bin&#8203;:/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:/home/runefro/bin
>     PERL\_BADLANG \(unset\)
>     SHELL=/bin/bash
> 

\-\- 
Lady Nancy Astor&#8203;: If you were my husband\, I would flavour your coffee
with poison\.
Churchill&#8203;: Madam \- if I were your husband\, I would drink it\.

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Nov 25\, 2002 at 12&#8203;:20&#8203;:23AM \+0000\, Dave Mitchell wrote&#8203;:
> The code below can reduced the simpler
> 
>     @&#8203;a = \(1\,2\,3\);
>     for \(sort\(@&#8203;a\)\) \{ 
>     @&#8203;a = \(\);
>     \}

Actually\, I've just noticed that this bug report is a duplicate of
\#18489\, and in that version\, Andreas reduced it to an almost identical
test case\!

\-\- 
You never really learn to swear until you learn to drive\.
